crates/core_simd/src/swizzle.rs

```diff
@@ -364,4 +364,131 @@ where
 
         (Even::swizzle2(self, other), Odd::swizzle2(self, other))
     }
+
+    /// Takes a slice of a vector to produce a shorter vector.
+    ///
+    /// This is equivalent to computing `&self[OFFSET..OFFSET+LEN]` on
+    /// the underlying array.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3, 4, 5, 6, 7]);
+    /// let y = x.slice::<2, 4>();
+    /// assert_eq!(y.to_array(), [2, 3, 4, 5]);
+    /// ```
+    ///
+    /// Will be rejected at compile time if `OFFSET + LEN > LANES`.
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn slice<const OFFSET: usize, const LEN: usize>(self) -> Simd<T, LEN>
+    where
+        LaneCount<LEN>: SupportedLaneCount,
+    {
+        const fn slice_index<const LEN: usize>(offset: usize, lanes: usize) -> [usize; LEN] {
+            assert!(offset + LEN <= lanes, "slice out of bounds");
+            let mut index = [0; LEN];
+            let mut i = 0;
+            while i < LEN {
+                index[i] = i + offset;
+                i += 1;
+            }
+            index
+        }
+        struct Slice<const OFFSET: usize>;
+        impl<const OFFSET: usize, const LEN: usize, const LANES: usize> Swizzle<LANES, LEN>
+            for Slice<OFFSET>
+        {
+            const INDEX: [usize; LEN] = slice_index::<LEN>(OFFSET, LANES);
+        }
+        Slice::<OFFSET>::swizzle(self)
+    }
+
+    /// Concatenates two vectors of equal length.
+    ///
+    /// Due to limitations in const generics, the length of the resulting vector cannot be inferred
+    /// from the input vectors. You must specify it explicitly.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3]);
+    /// let y = Simd::from_array([4, 5, 6, 7]);
+    /// let z = x.concat_to::<8>(y);
+    /// assert_eq!(z.to_array(), [0, 1, 2, 3, 4, 5, 6, 7]);
+    /// ```
+    ///
+    /// Will be rejected at compile time if `LANES * 2 != DOUBLE_LANES`.
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn concat_to<const DOUBLE_LANES: usize>(self, other: Self) -> Simd<T, DOUBLE_LANES>
```

Comment:
imho `concat` should support concatenating vectors of arbitrary length, e.g. `Simd::concat::<15>(my_i32x5, my_i32x10)`

Reply:
untested code for concatenating arbitrary vector lengths:
```rust
impl<T, const LANES: usize> Simd<T, LANES>
where
    T: SimdElement,
    LaneCount<LANES>: SupportedLaneCount,
{
    /// use like `a.concat::<15, _>(b)`
    pub fn concat<const OUTPUT_LANES: usize, const LANES2: usize>(self, other: Simd<T, LANES2>) -> Simd<T, OUTPUT_LANES>
    where
        LaneCount<OUTPUT_LANES>: SupportedLaneCount,
        LaneCount<LANES2>: SupportedLaneCount,
    {
        struct Extend;
        impl<const I: usize, const O: usize> Swizzle<I, O> for Extend {
            const INDEX: [usize; O] = {
                assert!(I <= O);
                let mut retval = [0; O];
                let mut i = 0;
                while i < I {
                    retval[i] = i;
                }
                retval
            };
        }
        struct Concat<const A: usize, const B: usize, const Y: usize>;
        impl<const A: usize, const B: usize, const Y: usize> Swizzle2<Y, Y> for Concat<A, B, Y> {
            const INDEX: [Which; Y] = {
                assert!(A + B == Y, "concat: OUTPUT_LANES must be the sum of all input lane counts");
                let mut retval = [Which::First(0); Y];
                let mut i = 0;
                while i < Y {
                    if i < A {
                        retval[i] = Which::First(i);
                    } else {
                        retval[i] = Which::Second(i - A);
                    }
                }
                retval
            };
        }
        Concat::<LANES, LANES2, OUTPUT_LANES>::swizzle2(Extend::swizzle(self), Extend::swizzle(other))
    }
}
```

Reply:
Sure, but does it work with type inference?

Reply:
> Sure, but does it work with type inference?

about as well as the size-doubling `concat` imho -- you have to specify the output size.

Reply:
> Given how `Swizzle2` works right now, arbitrary concats are not possible.

they're totally possible, just not using only one `Swizzle2` operation. the example code I posted uses separate swizzles to extend the inputs to the same length so `Swizzle2` can be used to combine them.

> The `Swizzle2` trait assumes that both of the two input vectors have the same length. We could potentially change that if you like, but I don't know all of the consequences of doing so.

`swizzle2` matches LLVM IR's `shuffle` instruction, so if we extend it to accept inputs of different length we would have to generate more complex IR -- basically matching the `concat` demo code I posted.

Reply:
I see. I misread your demo code. Thanks, done!

The new variant is slightly worse for type inference, because you can't write `x.concat::<4, _>(y)` without `generic_arg_infer` (https://github.com/rust-lang/rust/issues/85077); you have to write `x.concat::<4, 2>(y)` or whatever. That said, you can still constrain the `OUTPUT_LANES` by context, e.g. by giving a type signature to the result: `let z: u32x4 = x.concat(y);`.